include/power_grid_model/math_solver/iterative_pf_solver.hpp

```diff
@@ -0,0 +1,170 @@
+        output.shunt = y_bus.calculate_shunt_flow(output.u);
+
+        // prepare source and load gen
+        output.source.resize(source_bus_indptr_->back());
+        output.load_gen.resize(load_gen_bus_indptr_->back());
+        // loop all bus
+        for (Idx bus = 0; bus != n_bus_; ++bus) {
+            // source
+            for (Idx source = (*source_bus_indptr_)[bus]; source != (*source_bus_indptr_)[bus + 1]; ++source) {
+                ComplexValue<sym> const u_ref{input.source[source]};
+                ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
+                output.source[source].i = dot(y_ref, u_ref - output.u[bus]);
+                output.source[source].s = output.u[bus] * conj(output.source[source].i);
+            }
+
+            // load_gen
+            for (Idx load_gen = (*load_gen_bus_indptr_)[bus]; load_gen != (*load_gen_bus_indptr_)[bus + 1];
+                 ++load_gen) {
+                LoadGenType const type = (*load_gen_type_)[load_gen];
+                switch (type) {
+                    case LoadGenType::const_pq:
+                        // always same power
+                        output.load_gen[load_gen].s = input.s_injection[load_gen];
+                        break;
+                    case LoadGenType::const_y:
+                        // power is quadratic relation to voltage
+                        output.load_gen[load_gen].s =
+                            input.s_injection[load_gen] * cabs(output.u[bus]) * cabs(output.u[bus]);
+                        break;
+                    case LoadGenType::const_i:
+                        // power is linear relation to voltage
+                        output.load_gen[load_gen].s = input.s_injection[load_gen] * cabs(output.u[bus]);
+                        break;
+                    default:
+                        throw MissingCaseForEnumError("Power injection", type);
+                }
+                output.load_gen[load_gen].i = conj(output.load_gen[load_gen].s / output.u[bus]);
+            }
+        }
+    }
+
+   protected:
+    Idx n_bus_;
+    std::shared_ptr<DoubleVector const> phase_shift_;
+    std::shared_ptr<IdxVector const> load_gen_bus_indptr_;
+    std::shared_ptr<IdxVector const> source_bus_indptr_;
+    std::shared_ptr<std::vector<LoadGenType> const> load_gen_type_;
+
+   private:
+    IterativePFSolver(YBus<sym> const& y_bus, std::shared_ptr<MathModelTopology const> const& topo_ptr)
+        : n_bus_{y_bus.size()},
+          phase_shift_{topo_ptr, &topo_ptr->phase_shift},
+          load_gen_bus_indptr_{topo_ptr, &topo_ptr->load_gen_bus_indptr},
+          source_bus_indptr_{topo_ptr, &topo_ptr->source_bus_indptr},
+          load_gen_type_{topo_ptr, &topo_ptr->load_gen_type} {
+    }
+    friend DerivedSolver;
+};
+
+}  // namespace math_model_impl
+
+template <bool sym, typename DerivedSolver>
+using IterativePFSolver = math_model_impl::IterativePFSolver<sym, DerivedSolver>;
```

Comment:
I don't think you need this. The derived class will be implemented in the namespace `math_model_impl` anyway